test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc

```diff
@@ -0,0 +1,138 @@
+#include "envoy/config/core/v3/base.pb.h"
+#include "envoy/config/core/v3/proxy_protocol.pb.h"
+#include "envoy/extensions/transport_sockets/proxy_protocol/v3/upstream_proxy_protocol.pb.h"
+
+#include "test/integration/integration.h"
+
+namespace Envoy {
+namespace {
+
+class ProxyProtocolIntegrationTest : public testing::TestWithParam<Network::Address::IpVersion>,
+                                     public BaseIntegrationTest {
+public:
+  ProxyProtocolIntegrationTest()
+      : BaseIntegrationTest(GetParam(), ConfigHelper::tcpProxyConfig()) {}
+
+  void TearDown() override {
+    test_server_.reset();
+    fake_upstreams_.clear();
+  }
+
+  void setVersion(envoy::config::core::v3::ProxyProtocolConfig_Version version) {
+    version_ = version;
+  }
+
+  void initialize() override {
+    config_helper_.addConfigModifier([this](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+      auto* transport_socket =
+          bootstrap.mutable_static_resources()->mutable_clusters(0)->mutable_transport_socket();
+      transport_socket->set_name("envoy.transport_sockets.upstream_proxy_protocol");
+      envoy::config::core::v3::TransportSocket raw_transport_socket;
+      raw_transport_socket.set_name("envoy.transport_sockets.raw_buffer");
```

Comment:
The header should be sent first unencrypted.

Reply:
I was referring to the case where the proxy is configured to use a point-to-point TLS connection from the proxy to the upstream.

I agree that the usual use of TCP proxy with proxy protocol is something along the lines of client sending SSL to the proxy, proxy adding PROXY protocol header and forwarding the bytes from the client which may be wrapped in SSL.

Reply:
Sorry if I'm misunderstanding what you're getting at, but the handshake does not happen before the header is sent in this implementation. So if the proxy -> upstream socket setup is the proxy proto socket wrapping a tls socket, the header will be sent first unencrypted when a new connection is made, then the handshake will happen and the proxied data will be sent encrypted.

Reply:
ok, I looked at how you actually write the header.  You access the IoHandle directly instead of writing via the the wrapped transport_socket_->write method, which I think does result in the behavior you describe.  I think it does assume that doWrite is called before doRead, the SslSocket class calls doHandshake from doRead and said call can end up writing the first bytes of the handshake.

https://github.com/envoyproxy/envoy/blob/1df7242057f3aba3b52d8fcbf8f5169b6a8e8f22/source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc#L86

Reply:
That's true. Is there a case where `doRead` would get called first for an upstream connection?

Reply:
It really depends on how the kernel delivers events and if it would ever deliver spurious events or envoy inject some fake read events for whatever reason.  I think the failure mode that I was describing is unlikely to happen, seems fine to proceed with your version.